### PR TITLE
feat(broccoli): Add naive caching strategy for Broccoli.

### PR DESCRIPTION
**Pull request by amiller-gh**
- Aggregator Broc plugin now lives in Broccoli package.
 - Use fs-tree-diff to better keep input and output dirs in sync with minimal changes.
 - Don't do any work if input directory has not changed.
 - If input directory has changed, we still re-build the world. This can be improved.
 - Add more robust tests for Analyze and Aggregate plugins with broccoli-test-helper.

### CODE REVIEW

File: packages/@css-blocks/broccoli/src/Aggregate.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+
```
Comment (amiller-gh) [inline packages/@css-blocks/broccoli/src/Aggregate.ts]: File moved from `@css-blocks/ember-cli` and augmented with simple caching.

File: packages/@css-blocks/broccoli/src/Transport.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { Analyzer, Block, StyleMapping } from "@css-blocks/core";
```
Comment (amiller-gh) [inline packages/@css-blocks/broccoli/src/Transport.ts]: Moved in here from `@css-blocks/ember-cli`.

File: packages/@css-blocks/ember-cli/index.js [hunk 3]
```diff
@@ -148,7 +79,10 @@ module.exports = {
     // For Ember
     registry.add("htmlbars-ast-plugin", {
       name: "css-blocks-htmlbars",
-      plugin: this.astPlugin.bind(this)
+      plugin: this.astPlugin.bind(this),
+      // cacheKey: () => {
```
Comment (amiller-gh) [inline packages/@css-blocks/ember-cli/index.js]: @stefanpenner, going to need a little help understanding `cacheKey` here so I can cache results from my htmlbars plugin. For the naive approach, we ideally bust the cache whenever the class mapping is swapped out for a new one.

File: packages/@css-blocks/broccoli/src/Aggregate.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    this.out = out;
+  }
+
+  build() {
+    let output = this.outputPath;
+    let input = this.inputPaths[0];
+    let { id, css } = this.transport;
+
+    // Test if anything has changed since last time. If not, skip trying to update tree.
+    let newFsTree = new FSTree({ entries: walkSync.entries(input) });
```
Comment (stefanpenner) [inline packages/@css-blocks/broccoli/src/Aggregate.ts]: nit: `FSTree.fromEntries(walkSync.entries(input))`

File: packages/@css-blocks/broccoli/src/Aggregate.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      name: "broccoli-css-blocks-aggregate",
+      persistentOutput: true,
+    });
+    this.transport = transport;
+    this.out = out;
+  }
+
+  build() {
+    let output = this.outputPath;
+    let input = this.inputPaths[0];
```
Comment (stefanpenner) [inline packages/@css-blocks/broccoli/src/Aggregate.ts]: mind briefly describing what the expected inputPaths[0] is here (in this thread is fine)? Is it the full app, is it templates, is it...

File: packages/@css-blocks/broccoli/src/Analyze.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+  async build() {
+    let input = this.inputPaths[0];
+    let output = this.outputPath;
+    let options = this.analyzer.cssBlocksOptions;
+    let blockCompiler = new BlockCompiler(postcss, options);
+    let optimizer = new Optimizer(this.optimizationOptions, this.analyzer.optimizationOptions);
+
+    // Test if anything has changed since last time. If not, skip all analysis work.
+    let newFsTree = new FSTree({ entries: walkSync.entries(input) });
```
Comment (stefanpenner) [inline packages/@css-blocks/broccoli/src/Analyze.ts]: nit: `FSTree.fromEntries(walkSync.entries(input))`

File: packages/@css-blocks/broccoli/src/utils.ts [hunk 7]
```diff
@@ -3,11 +3,39 @@
...
+declare function require(id: string): any;
 
+export const BroccoliPlugin: BroccoliPlugin.Static = require("broccoli-plugin");
+export const symlinkOrCopy: (a: string, b: string) => void = require("symlink-or-copy").sync;
+export const FSTree: FSTree.Static = require("fs-tree-diff");
+export const FSEntry: FSTree.Entry = require("fs-tree-diff/lib/entry.js");
 export const walkSync: WalkSync = require("walk-sync");
 
-declare function require(id: string): any;
+export namespace FSTree {
```
Comment (stefanpenner) [inline packages/@css-blocks/broccoli/src/utils.ts]: If you have cycles to submit this to `fs-tree-sync` library itself, I'll gladly accept the typings.

Comment (amiller-gh) [reply packages/@css-blocks/broccoli/src/utils.ts]: Happily! 🎉 
https://github.com/stefanpenner/fs-tree-diff/pull/82

Comment (amiller-gh) [reply packages/@css-blocks/broccoli/src/utils.ts]: When you cut a new version of `fs-tree-diff` with the types I'll update the dep in this PR.

File: packages/@css-blocks/broccoli/src/Aggregate.ts (hunk 5 above)
Comment (amiller-gh) [reply packages/@css-blocks/broccoli/src/Aggregate.ts]: Is in jsdoc comments now 👍

File: packages/@css-blocks/ember-cli/index.js (hunk 3 above)
Comment (amiller-gh) [reply packages/@css-blocks/ember-cli/index.js]: @stefanpenner bump 😉

Comment (amiller-gh) [reply packages/@css-blocks/ember-cli/index.js]: Going to pull in for now and file a ticket for ember-cli-htmlbars plugin caching. After digging for a while I don't see how this is possible with the htmlbars plugin cache as is...

**Pull request merged**